File: app/src/main/java/igrek/todotree/service/calc/NumericAdder.java
package igrek.todotree.service.calc;


import java.math.BigDecimal;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import igrek.todotree.domain.treeitem.AbstractTreeItem;

public class NumericAdder {
	
	public BigDecimal calculateSum(Set<Integer> selectedPositions, AbstractTreeItem currentItem) throws NumberFormatException {
		BigDecimal sum = new BigDecimal(0);
		for (Integer selectedPos : selectedPositions) {
			AbstractTreeItem selectedItem = currentItem.getChild(selectedPos);
			BigDecimal itemValue = getItemNumericValue(selectedItem.getDisplayName());
			if (itemValue != null) {
				sum = sum.add(itemValue);
			}
		}
		return sum;
	}
	
	private BigDecimal getItemNumericValue(String content) throws NumberFormatException {
		
		String valueStr = null;
		
		content = content.replace(',', '.');
		
		Pattern pattern = Pattern.compile("^(-? ?\\d+(\\.\\d+)?)(.*?)$");
		Matcher matcher = pattern.matcher(content);
		if (matcher.matches()) {
			valueStr = matcher.group(1);
			valueStr = valueStr.replaceAll(" ", "");
		} else {
			pattern = Pattern.compile("^(.*?)(-? ?\\d+(\\.\\d+)?)$");
			matcher = pattern.matcher(content);
			if (matcher.matches()) {
				valueStr = matcher.group(2);
				valueStr = valueStr.replaceAll(" ", "");
			}
		}
		
		if (valueStr != null) {
			try {
				return new BigDecimal(valueStr);
			} catch (NumberFormatException e) {
				throw new NumberFormatException("Invalid number format:\n" + valueStr);
			}
		} else {
			throw new NumberFormatException("No number value:\n" + content);
		}
	}
}

File: app/src/main/java/igrek/todotree/service/calc/NumericAdder.kt
package igrek.todotree.service.calc

import igrek.todotree.domain.treeitem.AbstractTreeItem
import java.math.BigDecimal
import java.util.regex.Pattern

class NumericAdder {

    @Throws(NumberFormatException::class)
    fun calculateSum(selectedPositions: Set<Int>, currentItem: AbstractTreeItem): BigDecimal {
        var sum = BigDecimal(0)
        for (selectedPos in selectedPositions) {
            val selectedItem = currentItem.getChild(selectedPos)
            val itemValue = getItemNumericValue(selectedItem.displayName)
            sum = sum.add(itemValue)
        }
        return sum
    }

    @Throws(NumberFormatException::class)
    private fun getItemNumericValue(content: String): BigDecimal {
        var content = content
        var valueStr: String? = null
        content = content.replace(',', '.')
        var pattern = Pattern.compile("^(-? ?\\d+(\\.\\d+)?)(.*?)$")
        var matcher = pattern.matcher(content)
        if (matcher.matches()) {
            valueStr = matcher.group(1)
            valueStr = valueStr.replace(" ".toRegex(), "")
        } else {
            pattern = Pattern.compile("^(.*?)(-? ?\\d+(\\.\\d+)?)$")
            matcher = pattern.matcher(content)
            if (matcher.matches()) {
                valueStr = matcher.group(2)
                valueStr = valueStr.replace(" ".toRegex(), "")
            }
        }
        return if (valueStr != null) {
            try {
                BigDecimal(valueStr)
            } catch (e: NumberFormatException) {
                throw NumberFormatException("Invalid number format:\n$valueStr")
            }
        } else {
            throw NumberFormatException("No number value:\n$content")
        }
    }
}